src/sync/locks.rs

```diff
@@ -0,0 +1,198 @@
+use super::*;
+use core::{
+  cell::UnsafeCell,
+  mem::MaybeUninit,
+  ops::{Deref, DerefMut},
+  ptr,
+  sync::atomic::{compiler_fence, Ordering},
+};
+
+#[inline(never)]
+fn already_locked() -> ! {
+  panic!("This lock has already been locked by another thread.")
+}
+
+/// A mutex that prevents code from running in both an IRQ and normal code at
+/// the same time.
+///
+/// Note that this does not support blocking like a typical mutex, and instead
+/// mainly exists for memory safety reasons.
+pub struct RawMutex(Static<bool>);
+impl RawMutex {
+  /// Creates a new lock.
+  pub const fn new() -> Self {
+    RawMutex(Static::new(false))
+  }
+
+  /// Locks the mutex and returns whether a lock was successfully acquired.
+  fn raw_lock(&self) -> bool {
+    if self.0.replace(true) {
+      // value was already true, opps.
+      false
+    } else {
+      // prevent any weird reordering, and continue
+      compiler_fence(Ordering::Acquire);
+      true
+    }
+  }
+
+  /// Unlocks the mutex.
+  fn raw_unlock(&self) {
+    compiler_fence(Ordering::Release);
+    if !self.0.replace(false) {
```

Comment:
Note that if it was ever already false, it's not an "already locked" case so much as "there's a fundamental bug in the Mutex algorithm (or with the primitives used to implement it).

Reply:
Panic is effectively halt/abort here right? (Since this indicates that cross-thread memory corruption has already occurred and everything is ruined forever)

Reply:
yeah panic halts the cpu

Reply:
Yep. It's technically up to the user to write the panic hook, but it's expected that `panic = "abort"` regardless, even if the hook doesn't halt itself.

Reply:
Seems fine. The way its written this can only happen if there's some sort of buffer run or other UB anyway.